### PR TITLE
Use ID for Person equality #135

### LINKED ISSUE #135: Use ID for Person equality
Also add person retrieving (as read-only Person superinterface) by ID capabilities to `ModelManager` so external components cannot change the in-model Person objects.

Model/Remote should control IDs of newly added Persons.

### PR DESCRIPTION
**Pull request by YijinL**
Resolves #135 
Prerequisite for #196 completion.

### CODE REVIEW

File: src/main/java/address/model/datatypes/person/ReadOnlyPerson.java [hunk 1]
```diff
@@ -2,34 +2,104 @@
...
+        }
+        return changed;
+    }
+
+    /**
+     * Remote-assigned (canonical) ids are positive integers.
+     * Locally-assigned temporary ids are negative integers.
+     * 0 is reserved for ID-less Person data containers.
+     */
+    int getID();
```
Comment (damithc) [inline src/main/java/address/model/datatypes/person/ReadOnlyPerson.java]: `getId()`?

File: src/main/java/address/model/ModelManager.java [hunk 2]
```diff
@@ -147,27 +143,16 @@ public void clearModel() {
...
-        final Person toAdd = new Person(personToAdd);
-        if (backingPersonList().contains(toAdd)) {
-            throw new DuplicatePersonException(personToAdd);
-        }
+    public synchronized void addPerson(ReadOnlyPerson data) throws DuplicatePersonException {
+        Person toAdd;
+        do { // make sure no id clashes.
+            toAdd = new Person(Math.abs(UUID.randomUUID().hashCode()));
+        } while (backingPersonList().contains(toAdd));
+        toAdd.update(data);
```
Comment (yl-coder) [inline src/main/java/address/model/ModelManager.java]: Scalability concerns?

**Comment by damithc** [comment]
No more comments from me

File: src/main/java/address/model/datatypes/person/ReadOnlyPerson.java [hunk 3]
```diff
@@ -2,34 +2,104 @@
...
+        final Set<Integer> idSet = new HashSet<>(idsToRemove);
+        final Iterator<? extends ReadOnlyPerson> iter = col.iterator();
+        boolean changed = false;
+        while (iter.hasNext()) {
+            if (idSet.contains(iter.next().getID())) {
+                iter.remove();
+                changed = true;
+            }
+        }
+        return changed;
```
Comment (yl-coder) [inline src/main/java/address/model/datatypes/person/ReadOnlyPerson.java]: Will stream be better for this?
var result = stream.filter(...
col.removeAll(result)

File: src/main/java/address/model/ModelManager.java (hunk 2 above)
Comment (YijinL) [inline src/main/java/address/model/ModelManager.java]: This is a temporary workaround for assigning unique ids, will be replaced by a proper id assigning implementation in the cloud component

File: src/main/java/address/model/datatypes/person/ReadOnlyPerson.java (hunk 3 above)
Comment (YijinL) [inline src/main/java/address/model/datatypes/person/ReadOnlyPerson.java]: this method allows different ReadOnlyPersons implementations to be used as deletion keys (using id). `.Collection.removeAll` checks the implementation's `.equals` method

**Pull request merged**